libs/hooks/debounce/src/lib/debounce-hook.ts

```
@@ -0,0 +1,190 @@
+    });
+  }
+
+  before(hookContext: HookContext, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'before',
+      () => this.options?.beforeCacheKeySupplier?.(hookContext.flagKey, hookContext.context),
+      () => this.innerHook?.before?.(hookContext, hookHints),
+    );
+  }
+
+  after(hookContext: HookContext, evaluationDetails: EvaluationDetails<T>, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'after',
+      () => this.options?.afterCacheKeySupplier?.(hookContext.flagKey, hookContext.context, evaluationDetails),
+      () => this.innerHook?.after?.(hookContext, evaluationDetails, hookHints),
+    );
+  }
+
+  error(hookContext: HookContext, err: unknown, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'error',
+      () => this.options?.errorCacheKeySupplier?.(hookContext.flagKey, hookContext.context, err),
+      () => this.innerHook?.error?.(hookContext, err, hookHints),
+    );
+  }
+
+  finally(hookContext: HookContext, evaluationDetails: EvaluationDetails<T>, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'finally',
+      () => this.options?.finallyCacheKeySupplier?.(hookContext.flagKey, hookContext.context, evaluationDetails),
+      () => this.innerHook?.finally?.(hookContext, evaluationDetails, hookHints),
+    );
+  }
+
+  private maybeSkipAndCache(
+    stage: 'before' | 'after' | 'error' | 'finally',
+    keyGenCallback: () => string | null | undefined,
+    hookCallback: () => void,
+  ) {
+    // the cache key is a concatenation of the result of calling keyGenCallback and the stage
+    const dynamicKey = keyGenCallback();
+
+    // if the keyGenCallback returns nothing, we don't do any caching
+    if (dynamicKey) {
+      const cacheKeySuffix = stage;
+      const cacheKey = `${dynamicKey}::${cacheKeySuffix}`;
+      const got = this.cache.get(cacheKey);
+
+      if (got) {
+        // throw cached errors
+        if (got instanceof CachedError) {
+          throw got;
```

Won't this throw the wrapped error? Should we throw the underlying error instead?